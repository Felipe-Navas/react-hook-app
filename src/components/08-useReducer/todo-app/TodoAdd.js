import React from 'react';
import { useForm } from '../../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
  const [{ description }, handleInputChange, reset] = useForm({
    description: '',
  });

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }
    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Add TODO</h4>
      <hr />
      <form onSubmit={handleSubmit}>
        <input
          className="form-control"
          type="text"
          name="description"
          placeholder="Enter a todo description"
          autoComplete="off"
          onChange={handleInputChange}
          value={description}
        />
        <div className="d-grid gap-2">
          <button className="btn btn-success mt-1" type="submit">
            Add Todo
          </button>
        </div>
      </form>
    </>
  );
};
