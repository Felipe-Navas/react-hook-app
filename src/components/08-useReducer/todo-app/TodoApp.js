import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { useForm } from '../../../hooks/useForm';
import { TodoList } from './TodoList';
import './styles.css';

const init = () => {
  // return [
  //   {
  //     id: new Date().getTime(),
  //     desc: 'Deploy an application',
  //     done: false,
  //   },
  // ];

  return JSON.parse(localStorage.getItem('Todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const [{ description }, handleInputChange, reset] = useForm({
    description: '',
  });

  useEffect(() => {
    localStorage.setItem('Todos', JSON.stringify(todos));
  }, [todos]);

  const handleDelete = (todoId) => {
    const action = {
      type: 'delete',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleToggle = (todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId,
    };
    dispatch(action);
  };

  const handleSubmit = (e) => {
    e.preventDefault();

    if (description.trim().length <= 1) {
      return;
    }

    const newTodo = {
      id: new Date().getTime(),
      desc: description,
      done: false,
    };
    const action = {
      type: 'add',
      payload: newTodo,
    };
    dispatch(action);
    reset();
  };

  return (
    <div>
      <h1>TodoApp ({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToggle={handleToggle}
          />
        </div>
        <div className="col-5">
          <h4>Add TODO</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              className="form-control"
              type="text"
              name="description"
              placeholder="Enter a todo description"
              autoComplete="off"
              onChange={handleInputChange}
              value={description}
            />
            <div className="d-grid gap-2">
              <button className="btn btn-success mt-1" type="submit">
                Add Todo
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
