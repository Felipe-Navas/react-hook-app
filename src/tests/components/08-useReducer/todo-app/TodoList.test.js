import React from 'react';
import { shallow } from 'enzyme';
import { TodoList } from '../../../../components/08-useReducer/todo-app/TodoList';
import { demoTodos } from '../../../fixtures/demoTodo';

describe('Testing the <TodoList /> component', () => {
  const handleDelete = jest.fn();
  const handleToggle = jest.fn();

  const wrapper = shallow(
    <TodoList
      todos={demoTodos}
      handleDelete={handleDelete}
      handleToggle={handleToggle}
    />
  );
  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should have the same lenght of elements for the <TodoListItem /> components', () => {
    expect(wrapper.find('TodoListItem').length).toBe(demoTodos.length);
    expect(wrapper.find('TodoListItem').at(0).prop('handleDelete')).toEqual(
      expect.any(Function)
    );
    expect(wrapper.find('TodoListItem').at(0).prop('handleToggle')).toEqual(
      expect.any(Function)
    );
  });
});
