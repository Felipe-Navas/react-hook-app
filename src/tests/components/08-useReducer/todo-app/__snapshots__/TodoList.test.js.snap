// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the <TodoList /> component should render correctly 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    indice={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    indice={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Node",
        "done": false,
        "id": 2,
      }
    }
  />
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    indice={2}
    key="3"
    todo={
      Object {
        "desc": "Aprender GraphQL",
        "done": false,
        "id": 3,
      }
    }
  />
</ul>
`;
